Add regex patterns for CnsRegisterVolume Spec params

BaluDontu opened the pull request:
**What this PR does / why we need it**:
Currently CnsRegisterVolume Spec params doesn't have any validation rules. Adding validation rules for the params will esnure the CRD instance is not created in first place instead of creating it and failing it later by the CnsRegisterVolume controller.

Using the following regex patterns for the following params
1. pvcName - Following the rules for the resource name defined in https://kubernetes.io/docs/concepts/overview/working-with-objects/names/
2. volumeID - UUID format
3. diskUrlPath - url format

```release-note
Add regex patterns for CnsRegisterVolume Spec params
```

**Testing done**
"cnsregistervolumes.cns.vmware.com" CRD is updated on deploying new CSI driver.

```
2020-10-16T22:33:16.870Z        INFO    kubernetes/kubernetes.go:345    "cnsregistervolumes.cns.vmware.com" CRD updated successfully    {"TraceId": "6a68b934-b973-43c8-a4e8-94dade4de261"}
```

1. pvcName is not in valid format
```
root@42168c19636add845e5e2e951ca754d8 [ ~ ]# cat ll.yaml 
apiVersion: cns.vmware.com/v1alpha1
kind: CnsRegisterVolume
metadata:
  name: import-mongodb-3
spec:
  pvcName: Mongo-pvc
  volumeID: fe1efbd0-79f7-4a2f-a8a9-69852e62225a
  accessMode: ReadWriteOnce

root@42168c19636add845e5e2e951ca754d8 [ ~ ]# kubectl create -f ll.yaml 
The CnsRegisterVolume "import-mongodb-3" is invalid: spec.pvcName: Invalid value: "": spec.pvcName in body should match '^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$'
```

2. volumeID is not in proper format
```
root@42168c19636add845e5e2e951ca754d8 [ ~ ]# cat ll.yaml 
apiVersion: cns.vmware.com/v1alpha1
kind: CnsRegisterVolume
metadata:
  name: import-mongodb-3
spec:
  pvcName: mongo-pvc
  volumeID: fe1efbd0ddddddddd-79f7-4a2f-a8a9-69852e62225a
  accessMode: ReadWriteOnce

root@42168c19636add845e5e2e951ca754d8 [ ~ ]# kubectl create -f ll.yaml 
The CnsRegisterVolume "import-mongodb-3" is invalid: spec.volumeID: Invalid value: "": spec.volumeID in body should match '^[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}$'
```

cc @divyenpatel @chethanv28 @SandeepPissay

BaluDontu commented:
/jtest wcp

svcbot-qecnsdp commented:
Started WCP block pipeline...

SandeepPissay commented:
Looks good to me. Lets wait for E2E test runs.

/approve

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
Ran 29 of 106 Specs in 6567.653 seconds
FAIL! -- 1 Passed | 28 Failed | 0 Pending | 77 Skipped
--- FAIL: TestE2E (6567.68s)
FAIL

Ginkgo ran 1 suite in 1h50m49.83096809s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/115/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 
Jenkins E2E Test Results: 
Ran 29 of 106 Specs in 5853.968 seconds
FAIL! -- 1 Passed | 28 Failed | 0 Pending | 77 Skipped
--- FAIL: TestE2E (5854.00s)
FAIL

Ginkgo ran 1 suite in 1h39m1.348126962s
Test Suite Failed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD@2/Results/117/vsphere-csi-driver`

```

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

BaluDontu commented:
/jtest wcp

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

divyenpatel commented on pkg/apis/cnsoperator/config/cnsregistervolume_crd.yaml:
```diff
@@ -33,10 +33,12 @@ spec:
               description: VolumeID indicates an existing vsphere volume to be imported
                 into Project Pacific cluster
               type: string
+              pattern: '^[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}$'
```
I think we do not have FCD with capital case chars.

BaluDontu replied:
@divyenpatel is there any documentation which says that? For now, I have used UUID format. 
@SandeepPissay can you confirm on this as well?

SandeepPissay replied:
I would suggest to not do FCD specific validation here as we need to support the same API for file volumes as well.

BaluDontu replied:
makes sense.

SandeepPissay replied:
Can you explain this regex? What is allowed?

divyenpatel replied:
> @divyenpatel is there any documentation which says that?

I do not have a link to documentation, but I have not seen volume ID in caps.

but I have one more question. Do we anticipate static file share volume provisioning using same CRD mechanism, then this pattern needs to be updated? as file share volume id is having some pre-fixes before UUID.

BaluDontu replied:
regex says 
8 chars of (number, letters lower and caps) - 4 chars of (number, letters lower and caps) - 4 chars of (number, letters lower and caps) - 12 chars of (number, letters lower and caps)

divyenpatel commented on pkg/apis/cnsoperator/config/cnsregistervolume_crd.yaml:
```diff
@@ -45,6 +47,7 @@ spec:
               description: DiskUrlPath is URL path to an existing block volume to be
                 imported into Project Pacific cluster.
               type: string
+              pattern: '^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$'
```
Can we change this to check the valid file path URL, not just validate any www URL?

for example: `(http[s]?:\/\/)?([^\/\s]+\/folder\/)(.*)`

divyenpatel reviewed (changes requested):
Can we enhance regex validation further?

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

SandeepPissay commented:
@BaluDontu I see that the current code does not update the CRD if the manifest file is edited. See https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/kubernetes/kubernetes.go#L329. So, dont you have to fix this part of the code as well?

BaluDontu commented:
@SandeepPissay Thanks for pointing this out. I have added a new commit which updates the CRD if CRD already exists. Looks like this piece of code was missed when porting the code from internal gitlab to github.

@divyenpatel @SandeepPissay Please take a look at latest commit that addresses your comments.

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

SandeepPissay commented on pkg/apis/cnsoperator/config/cnsregistervolume_crd.yaml:
```diff
@@ -45,6 +47,7 @@ spec:
               description: DiskUrlPath is URL path to an existing block volume to be
                 imported into Project Pacific cluster.
               type: string
+              pattern: '^(http[s]?:\/\/)?([^\/\s]+\/folder\/)(.*)$'
```
Can we add validation for accessMode?

BaluDontu replied:
We don't want to do that as these values are of type v1.PersistentVolumeAccessMode and shouldn't be hardcoded as patterns in yaml. 

We already have such validations for these in controller.

SandeepPissay replied:
What happens if I set accessMode as "abc"? Would the CR creation fail?

BaluDontu replied:
No, CRD will be created but reconciler will mark this failed.

SandeepPissay replied:
Isn't that too late? Ideally we should reject invalid CRs before even storing it in API server. Can we use `enum` to validate the allowed strings?

BaluDontu replied:
I don't want to hardcode strings here. What if PersistentVolumeAccessMode adds new type, we should modify our YAMLs for it. So, I decided to add that as part of code only.

SandeepPissay replied:
What happens when the code fails, does it get invoked again and again with the invalid filed? Would that mean the controller keeps continuously failing forever? If yes, thats a problem since this could be avoided by proper validation in the first place to reject CR creation.

SandeepPissay commented on pkg/kubernetes/kubernetes.go:
```diff
@@ -323,15 +323,27 @@ func createCustomResourceDefinition(ctx context.Context, crd *apiextensionsv1bet
+		// Update the existing CRD with new CRD
+		crd.Spec = newCrd.Spec
+		crd.Status = newCrd.Status
+		_, err = apiextensionsClientSet.ApiextensionsV1beta1().CustomResourceDefinitions().Update(ctx, crd, metav1.UpdateOptions{})
```
Question: Why dont we just update the CRD from the manifest directly?

BaluDontu replied:
No, we cannot do like that. The manifest spec doesn't have any resource version etc for it to be updated with directly.
We get below error if we do like that

```
Error initializing Cns Operator. Error: customresourcedefinitions.apiextensions.k8s.io "cnsnodevmattachments.cns.vmware.com" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update
```

So, we update the spec and status directly on existing crd.

SandeepPissay replied:
So, maybe we should look at how `kubectl apply -f <file_name>` works. I think this is fine too but it is not equivalent to the kubectl command that updates everything.

BaluDontu replied:
May be i will take it up as a follow up. I will leave it as is for now.

SandeepPissay replied:
Can you file an issue for this so that we can track this enhancement?

BaluDontu replied:
Why is this an issue? Not sure if the current code has any problem.

SandeepPissay replied:
As long as we are updating only the spec and status in the yaml then thats fine. Once we start updating some other new fields in this CRD, this code would not work and need to be revisited. I hope we do not revisit this.

divyenpatel commented:
@BaluDontu just one comment for the future use case. Rest looks good to me - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/427/files#r508776528

BaluDontu commented:
```
08:32:38  ------------------------------
08:32:38  SSSSS
08:32:38  Ran 33 of 148 Specs in 7958.240 seconds
08:32:38  SUCCESS! -- 33 Passed | 0 Failed | 0 Pending | 115 Skipped
08:32:38  PASS
08:32:38 
```

BaluDontu commented:
@divyenpatel I will update the regex when file volumes is supported for static provisioning on WCP SV.

divyenpatel commented:
/approve
/lgtm
